teps/0146-parameters-in-script.md

```diff
@@ -0,0 +1,605 @@
+
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone $(params.url)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      IFS='' read -d -r PARAM_URL /tekton/params/url
+      git clone ${PARAM_URL}
+```
+
+This approach is also under discussion in the context of changing the way `Result`
+values are stored in [TEP-0086][tep-0086-alt]. In that proposal a sidecar is
+responsible for fetching `Results` from previous `Tasks` and writing them to
+a location on disk for use as `Parameters`.
+
+##### Discussion
+
+This approach aligns well with the proposal in [TEP-0086][tep-0086-prop], and
+brings consistency in how we supply both `Parameters` and `Results` to `Steps`.
+
+###  Combine Elements of Both Explicit Environment Variable and Filesystem Projection Solutions
+
+Building on environment variable alternatives (1.0 & 1.1), and filsystem alternatives (2) Tekton Pipelines could offer an explicit option to the user whether to project as env (language agnostic OR shell format) or file.  There could be different env syntaxes for shell (most used in tasks) vs language agnostic env vars to make the option more UX friendly and backwards compatible for the majority of tasks (which use `sh` or `bash`).
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone $(params.url)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone ${$(params.url1.env)}
+      git clone $(params.url2.env_shell)
+      git clone $(params.url3.path)
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_URL1
+      value: $(params.url1)
+    script: |
+      git clone ${PARAM_URL1}
+      git clone ${PARAM_URL2}
+      IFS='' read -d -r PARAM_URL3 /tekton/params/url3
+      git clone ${PARAM_URL3}
+```
+
+####  Discussion
+
+Similar to the other options this mutates the spec, though does so at the user's
+discretion. This alternative could be an extension of 1.0 and could be implemented piecemeal with only a single implicit env format supported initially (`.env`) with additional implicit env formats and filesystem (`$(params.bar.path)`) support added later if needed.
+
+###  Reject `Parameters` in `Script` and provide instructions to fix
+
+During validation we could detect `Parameters` used in `Script` blocks and return
+a helpful message to the user that they need to rewrite their `Task` to move the
+`Parameters` into environment variables.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone $(params.url)
+```
+
+When the Tekton Pipelines Controller sees this `Script` it rejects the `Task` submission
+with an error, such as:
+
+```
+spec.steps[0].script: $(params.url) cannot be injected into a script, consider moving
+into an environment variable on the Step.
+```
```

Comment:
If we did fail validation, we would then basically not support them at all, which would be a backwards incompatible change. We could do that in `v2`, for `v1` we might do that as an opt-in feature or a warning system - even though validation in itself does not have any facility for deprecations or warnings that I'm aware of, so it's unclear how the warning could be delivered to the user

Reply:
+1, currently the idea for rolling this out would be to keep backwards compatibility with an opt-in feature-gate for cluster operators to remove this in v1 (if we agree this makes sense).  For the warnings, it wasn’t clear to me the de-facto way to do this in Tekton.  As stated, we could output some “WARN: Tekton is deprecating <...>”, see docs link here for more information on deprecation timelines how to migrate <doc-link AND/OR link to TBD bespoke `tkn` command for migration>” on creation of a Task but not sure how visible this might be to most end users.